.github/workflows/daily.yml

```diff
@@ -0,0 +1,380 @@
+# This workflow is a daily cron job.
+# As well as running our own tests,
+# we also run the tests of various third-party libraries that use us.
+# This helps us spot regressions early,
+# and helps flag when third-party libraries are making incorrect assumptions
+# that might cause them to break when we cut a new release.
+
+name: Daily tests
+
+on:
+  schedule:
+    - cron: "0 0 * * *"  # Midnight UTC
+  pull_request:
+    paths:
+      - ".github/workflows/daily.yml"
+  workflow_dispatch:
+
+permissions:
+  contents: read
+
+env:
+  PIP_DISABLE_PIP_VERSION_CHECK: 1
+  FORCE_COLOR: 1
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
+
+jobs:
+  typing-extensions:
+    name: Run our own tests
```

Comment:
We don't _have_ to run our own tests as part of this workflow; I can possibly take this out. My reasons for including our own tests were:
- I was conceptualising this as the comprehensive "test everything we can possibly think of" workflow, and including our own tests makes sense as part of that effort. But if we rename this workflow to "third-party tests", as you suggested above, then that framing doesn't work so well.
- Our own tests _are_ susceptible to be broken by one external party in particular: CPython. There's lots of things we re-export from CPython on later versions of Python, and our tests for these things could easily get accidentally broken by changes in a CPython patch release. It's true we run our own tests on every PR and push, but we also have "quiet periods" when there's not much activitiy in the repo, and our tests aren't run particularly frequently as a result.

> We should at least use a reusable workflow, so we don't duplicate this logic with all problems that this entails.

I'm not too familiar with reusable workflows, but I'm happy to look into that. I agree the duplication isn't ideal (though my general philosophy is also that it's best to keep the logic workflow files relatively simple, or you end up having to write tests for the tests!)

Reply:
Another way of reducing the duplication between the workflows would be to add a `cron` trigger to our current "test and lint" workflow, instead of including those tests as part of this workflow.

But also, it isn't essential to run our own tests as part of this workflow. Happy to do whatever you think best.

Reply:
Adding a cron trigger to our "ci" workflow that runs maybe 30 minutes before this might be the best compromise.

Reply:
Sure, I'll do that